hyperspy/learn/mva.py

```diff
@@ -1636,7 +1648,990 @@ def undo_treatments(self):
+                                     "cluster source signal size")
+            # if it's not already unfolded - unfold but remember the unfolding so
+            # can restore later
+            if not cluster_source.metadata._HyperSpy.Folding.unfolded:
+                cluster_source.unfolded4clustering=cluster_source.unfold()
+            data = cluster_source.data \
+                if cluster_source.axes_manager[0].index_in_array == 0 else cluster_source.data.T
+            toreturn = data[:,signal_mask][navigation_mask,:]
+        elif type(cluster_source) is str:
+            if cluster_source == "bss":
+                loadings = self.learning_results.bss_loadings
+                factors  = self.learning_results.bss_factors
+            else:
+                loadings = self.learning_results.loadings
+                factors  = self.learning_results.factors
+        
+            if reproject:
+                toreturn = factors[:, :number_of_components] \
+                    @ loadings[navigation_mask, :number_of_components].T          
+                toreturn = toreturn.T       
+            else:
+                toreturn = loadings[navigation_mask,:number_of_components] 
+
+        return toreturn
+
+        
+    def cluster_analysis(self,
+                         cluster_source,
+                         source_for_centers=None,
+                         center_signals_method="mean",
+                         preprocessing=None,
+                         preprocessing_kwargs={},
+                         number_of_components=None,
+                         navigation_mask=None,
+                         signal_mask=None,
+                         algorithm=None,
+                         return_info=False,
+                         **kwargs):
+        """Cluster analysis of a signal or decomposition results of a signal
+        Results are stored in `learning_results`.
+        
+        
+        Parameters
+        ----------
+        cluster_source : {"bss", "decomposition", "signal", BaseSignal}
+            If "bss" the blind source separation results are used
+            If "decomposition" the decomposition results are used 
+            if "signal" the signal data is used 
+            Note that using the signal or BaseSignal can be memory intensive 
+            and is only recommended if the Signal dimension is small
+            BaseSignal must have the same navigation dimensions as the signal.
+        source_for_centers : {None,"decomposition","bss","signal",BaseSignal},
+            default : None
+            If None the cluster_source is used
+            If "bss" the blind source separation results are used
+            If "decomposition" the decomposition results are used 
+            if "signal" the signal data is used 
+            BaseSignal must have the same navigation dimensions as the signal.            
+        center_signals_method : {"mean","median","closest"}
+            default : 'mean'
```

Comment:
I wonder if we should make "closest" the default because I think that it is more robust than the other methods, what do you think?

Also, how can we make it easier for the user to compute the centers using different methods without executing ``cluster_analysis`` multiple times? Options that come to my mind are:

1. Calcule the centers using all 3 methods and choose what to get/plot with a keyword in those methods.
2. Split the code to calculate the centers into its own method.

I have a slight preference for 1.